### PR TITLE
[lldb] Add format eFormatEnumWithValues to ensure raw enum value is always shown

### PR DESCRIPTION
**Pull request by DavidSpickett**
When an enum is used to represent certain data it can be useful to know
its name and the value of it. For instance, register fields are 
often set in source code as numbers, but in the debugger you'd like
to see the meaning as well.

(lldb) register read fpcr
    fpcr = 0x00000000
         = (... RMode = RN (0), ...)

Often you do just want the meaning but the value saves you having
to manually decode it if you want to confirm what your source code
has done, or try to replicate the current state in your source code.

This also works for bitfield like enums, with the added change
that if a bitfield like enum has the value 0, we will print 0 if
asked to always show a value. Normally we don't print a 0 there
because 0 means no flags are set.

I did not intend to make this new format available to the user,
but it ended up being so. So you can do: 

expression --format "enumeration with values" -- foo 

So testing is mainly from c++ but I have added a couple to the 
Python tests.

### CODE REVIEW

**Comment by Endilll** [comment]
I'm excited so see changes in this area!

> It is a subset of https://github.com/llvm/llvm-project/pull/69815, only for use by register printing (ping @Endilll).

Yeah, that effort is stalled because I don't see a clear path forward there, and I'm lacking energy to find and push for one. I guess I can try to piggyback on this PR if it goes in the direction I need (I hope you don't mind :).

Seriously, though, I'm not sure how your design with a flag orthogonal to format fares in the feedback I was given back then. It's also not clear how this new "presentation mode" is going to be controlled. While it's not in the scope of this PR, it would be nice if you can share a bigger picture.

Another thing this PR doesn't seem to touch is flag-like enums. Even if you don't want to add a numerical value there, because it'd be too much information to present, it still would be nice to see it explicitly considered.

**Comment by DavidSpickett** [comment]
> It's also not clear how this new "presentation mode" is going to be controlled. While it's not in the scope of this PR, it would be nice if you can share a bigger picture.

For me the picture is limited to register printing. I'll be setting this new option on the register type when printing (https://github.com/DavidSpickett/llvm-project/commit/24dbefaa14030646ba0871298989558221af1ba0#diff-18135f619417bbaa1ab0c181ce55b2c55681323c06e90fa1a3e16099c7176e21). So only from C++ for now.

> Another thing this PR doesn't seem to touch is flag-like enums. Even if you don't want to add a numerical value there, because it'd be too much information to present, it still would be nice to see it explicitly considered.

What is a flag like enum?

**Comment by Endilll** [comment]
> For me the picture is limited to register printing. I'll be setting this new option on the register type when printing (https://github.com/DavidSpickett/llvm-project/commit/24dbefaa14030646ba0871298989558221af1ba0#diff-18135f619417bbaa1ab0c181ce55b2c55681323c06e90fa1a3e16099c7176e21). So only from C++ for now.

So it's hard-coded, I see.

> What is a flag like enum?

`enum { a = 1, b = 2, c = 4, d = 8}` — this sort of enumeration. Sorry if my terminology was confusing. They are handled further in `DumpEnumValue()`, after the changes you've made here, and called bit-fields in the comments.

**Comment by DavidSpickett** [comment]
I understand, so the value is made from `a | b | c` for example. I didn't think of whether register fields could hit this path, but it's equivalent to a C enum they'll be using, so I think it could happen.

I'll test this, thanks for bringing it up.

**Comment by DavidSpickett** [comment]
I found a few things:
* Yes you can get these flags like enums with a register field as expected.
* Only C++ code can manipulate a `DumpValueObjectOptions` object, there's no Python binding at all.
* The only Python tests for it are things that call commands that then use the object e.g. watchpoints or evaluating enum expressions. They do not have the low level control like I am adding in this PR.

I'm not sure of the utility of flag like enums as register fields yet but whatever I come up with, the lack of tests bothers me. So I'm trying to find a way to test `DumpValueObjectOptions` in relative isolation.

**Review by clayborg — CHANGES_REQUESTED** [review]
So each `ValueObject` has the ability to show its value as an enumeration if its format is set to `eFormatEnum`. If the format is set to `eFormatHex`, `eFormatUnsigned`, or `eFormatSigned`, then we show the numeric value. 

Can you show some sample output for this? I would like the `const char *SBValue::GetValue()` function to obey the format that is selected and not return something that has both inside of it. It would be ok for the `const char *SBValue::GetSummary()` to return the enumeration if the format isn't set to `eFormatEnum`.

This makes me question the need for the `ValueObject::SetEnumsAlwaysShowValue()` and `ValueObject::GetEnumsAlwaysShowValue()` methods, the ivars they use and many of the other changes here.

**Comment by clayborg** [comment]
If your register has fields, you can set its format to be eFormatEnum by default. Each register defines how it gets displayed when we query the dynamic register information from the `lldb-server` or `debugserver`

**Comment by DavidSpickett** [comment]
>So each ValueObject has the ability to show its value as an enumeration if its format is set to eFormatEnum. If the format is set to eFormatHex, eFormatUnsigned, or eFormatSigned, then we show the numeric value.

Sure, the problem I have is that often with registers you'd want to see not just the name but the value. It's not 100% crucial for the feature but I found it useful in a previous debugger I used.

>Can you show some sample output for this?

I only have examples for registers right now, for example AArch64's mte_ctrl will look like (where TCF is an enum):
```
= (TAGS = 65535, TCF = TCF_SYNC (1), TAGGED_ADDR_ENABLE = 1)
```
So that users know what a value means (no looking in the manuals) and the value (in case they need to add to or confirm the number they see in their source code).

When I've added tests for the existing options, I'll update this PR with test cases that are not register specific.

**Comment by DavidSpickett** [comment]
@clayborg I have updated this to instead use a format, inspired by `eFormatBytesWithASCII`. Which is also one value printed twice. There are examples in the added c++ tests.

File: lldb/include/lldb/lldb-enumerations.h [hunk 1]
```diff
@@ -195,11 +195,15 @@ enum Format {
...
-  eFormatVoid,        ///< Do not print this
+                         ///< with file/line, symbol + offset, data, etc)
+  eFormatHexFloat,       ///< ISO C99 hex float string
+  eFormatInstruction,    ///< Disassemble an opcode
+  eFormatVoid,           ///< Do not print this
   eFormatUnicode8,
+  eFormatEnumWithValues, ///< Format as an enum but if the value matches one or
+                         ///< more enumerators, print the enumerator name and
+                         ///< value of those enumerators. For example "foo (1)"
+                         ///< instead of "foo".
```
Comment (DavidSpickett) [inline lldb/include/lldb/lldb-enumerations.h:206]: This enum is used in the API so I've added to the end of it.

**Comment by DavidSpickett** [comment]
@Endilll This being a format type now might mean you can use it from the formatters. I'm not familiar with what parts of the API you're using there.

**Comment by DavidSpickett** [comment]
@clayborg ping!

**Comment by DavidSpickett** [comment]
ping!

**Comment by clayborg** [comment]
Sorry for the delay, I am still wondering if we just want to have the value objects for enums return the text name as the value, and have the summary show the value as an appropriate integer if the format is eFormatEnum. When you display registers bitfields do you have a ValueObject? If so we can modify the `ValueObject::GetSummary()` to return the integer value, but only if we have the format set to `eFormatEnum`. The output for variables would be similar to what your tests expect, but we would just modify the summary string that we return and then we don't need the new format. 

One reason I like this way of doing things is if you call "valobj.GetValue()", and the format is set the enum with value, we will get "foo (1)" as the value back from the value object. I would rather call `valobj.GetValue()` and get `"foo"` and then call `valojb.GetSummary()` and get `"1"`, or of course you can always just call `valobj.GetValueAsUnsigned(...)` or `valobj.GetValueAsSigned(...)`. If we have special display code for the register bitfields, they can call these APIs when displaying register by default?

**Comment by DavidSpickett** [comment]
> When you display registers bitfields do you have a ValueObject?

Yes:
https://github.com/llvm/llvm-project/blob/6808e6c78edf63d2c0c9bd6fce168fe7e0a7d49f/lldb/source/Core/DumpRegisterValue.cpp#L47

I'm just calling `dump`, there's no register specific code there, ideally I wouldn't need any.

> If so we can modify the ValueObject::GetSummary() to return the integer value, but only if we have the format set to eFormatEnum. The output for variables would be similar to what your tests expect, but we would just modify the summary string that we return and then we don't need the new format.

I am not sure if a call to `dump` will end up using `GetSummary` or whether I'd need to loop over the children of the object myself, I'll look into it.

> One reason I like this way of doing things is if you call "valobj.GetValue()", and the format is set the enum with value, we will get "foo (1)" as the value back from the value object. I would rather call valobj.GetValue() and get "foo" and then call valojb.GetSummary() and get "1"

When you put it like API style calls like that, I agree. `foo` is a "value" in the sense you could write it in your C source, `foo (1)` is not, it is literally a summary, an overview of the value, not usable in code.

> If we have special display code for the register bitfields, they can call these APIs when displaying register by default?

By using the existing type infrastructure I avoided having to write printer code, so there isn't any right now.

You could sort of do this by having options that only C++ code can set and are only used by the register dump code, this is what the original version of this PR did (which you can see [here](https://github.com/DavidSpickett/llvm-project/commit/cb875625d7159857345e3da2b9aa4f2436134bee) to compare).

It's not great because it is this one off use case, but at least it's hidden from the command line and API users. Or I could add some kind of callback that only register dumping sets, that asks how to format the enum, but this is the same thing with different steps.

Could wait and see if registers need anything else unique, and once enough of that exists, invest in custom code. After all, users can `register info` to see the values for an enum, and only a minority will need them anyway.

...and an off the wall idea I just thought of. These register types are only for display, and we could generate different ones for expressions (https://discourse.llvm.org/t/rfc-register-fields-in-expressions/79424) if needed. So instead of using `enum` for the field why not a `union` of `enum` and `unsigned`.

The printed format might not be obvious to the reader but I'll give it a try, as it would sidestep the problems of the other solutions.

**Comment by DavidSpickett** [comment]
Custom printing code for just enums for registers would be possible but it means copying a lot of existing code. A callback reduces that to a copy of `DumpEnumValue` only. Or you can iterate the value object yourself but then you have to provide the surrounding format too for example the `{}` around a struct and so on.

I found I can modify the summary by changing `TypeFormatImpl_Format::FormatObject`, but this has the problem that it does not know if the enum value will match an enumerator ahead of time. So it works well for `<value> <unconditional extra stuff>`, but not if you only sometimes want the extra stuff. So another lookahead or callback needed here.

None of the options are great and this is not a required feature so I'm going to drop this for now. Maybe one of them will stand out when I look again later.

**Pull request closed without merging**